### ISSUE TITLE
Suspicious incorrect implementation of `lyapunov_from_data`.

### ISSUE THREAD

**Issue by onkyo14taro** [JuliaDynamics/ChaosTools.jl#241]
## Discrepancy in the definitions

I think there might be a mistake in the implementation of [`lyapunov_from_data`](https://juliadynamics.github.io/DynamicalSystems.jl/latest/chaos/lyapunovs/#Lyapunov-exponent-from-data).  

In [Kantz1994], which is referenced in the implementation, the maximum Lyapunov exponent is estimated as follows (Eq. (2.3)):
<img width="325" alt="Eq. (2.3)" src="https://user-images.githubusercontent.com/15100931/149173808-c4c481f0-4648-4932-aab0-d8666f88b4d1.png">

On the other hand, the current implementation of `lyapunov_from_data` is done with the following:
<img width="354" alt="lyapunov_from_data" src="https://user-images.githubusercontent.com/15100931/149062334-28910c8c-6f66-4bcc-91e9-7ae3b9e429b7.png">

The difference between the two equations is the position of the logarithm. 

## Notes

I don't have the [Skokos2016] on which the implementation of `lyapunov_from_data` is based. However, if [Skokos2016] is a development of [Kantz1994], then I think the formula for estimating the Lyapunov exponent itself is the same in [Skokos2016] and [Kantz1994].

## References

[Kants1994] H. Kantz, “A robust method to estimate the maximal Lyapunov exponent of a time series,” Phys. Lett. A, vol. 185, no. 1, pp. 77–87, 1994.
[Skokos2016] Skokos, C. H. et al., Chaos Detection and Predictability - Chapter 1 (section 1.3.2), Lecture Notes in Physics 915, Springer (2016)

**Comment by Datseris** [comment]
Thanks for your concerns. I think the second version is more accurate. Ultimately the MLE is the mean exponential growth of perturbations over a single trajectory. In numerical data the assumption is made that the dataset covers the chaotic attractor and we can use points on the dataset to represent various trajectories. Now the question is, what do you average over? You can average the distance growth of the trajectories, or you can average the lyapunov exponent of the trajectories. I think the latter is more correct. Actually, I don't just think this, this is an analytic result. This is exactly what the ergodic theorem says.

**Comment by Datseris** [comment]
If you are interested it should be super easy to make a second version of the current code that uses the first formula, so we can see how much error it creates!

**Comment by onkyo14taro** [comment]
@Datseris 

Thank you very much for your valuable comments! I am a beginner in this field, and your words have helped me learn. As you said, I now understand that the second equation is mathematically more appropriate.

On the other hand, I was getting `-Inf` results when using `lyapunov_from_data` for my audio data. I investigated the cause and found that it was due to a few elements with a logarithmic distance of `-Inf` (i.e., a distance of `0`).

It seems extremely unlikely that the two elements will match, but the possibility increases with coarsely quantized data (e.g., audio data recorded with 16-bit integer).

So I think it would be better to have an option in `lyapunov_from_data` to prevent `-Inf`. Specifically, I think it would be possible to solve this problem by allowing the user to set a minimum distance threshold (`0` by default). How about adding a minimum threshold keyword argument?
